### PR TITLE
fix(repository): handle conversion of property values to plain json

### PR DESCRIPTION
**Pull request by raymondfeng**
This PR adds the ability to recursively convert model instances to plain json object

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: packages/repository/src/model.ts [hunk 1]
```diff
@@ -170,7 +174,13 @@ export abstract class Model {
     if (options && options.ignoreUnknownProperties === false) {
       obj = {};
       for (const p in this) {
-        obj[p] = this[p];
+        let val = (this as AnyObject)[p];
+        if (val != null && typeof val.toObject === 'function') {
+          val = val.toObject();
+        } else if (val != null && typeof val.toJSON === 'function') {
```
Comment (b-admike) [inline packages/repository/src/model.ts]: I guess `toJSON` is our fallback from the `toObject` method? I.e. why do we want to have this `else if` block?

Comment (raymondfeng) [reply packages/repository/src/model.ts]: Use `toJSON()` as the fallback

File: packages/repository/src/model.ts [hunk 2]
```diff
@@ -156,7 +156,11 @@ export abstract class Model {
 
     for (const p in def.properties) {
       if (p in this) {
-        json[p] = (this as AnyObject)[p];
+        let val = (this as AnyObject)[p];
```
Comment (jannyHou) [inline packages/repository/src/model.ts]: @raymondfeng What is the difference between `toJSON` and `toObject` function? I couldn't find related document talking about this 2 methods, shall we add some doc to explain the difference and their usage?

Comment (raymondfeng) [reply packages/repository/src/model.ts]: `toObject()` accepts some options which `toJSON()` does not. I improved the test to show that.

Comment (jannyHou) [reply packages/repository/src/model.ts]: thanks got it.

**Pull request merged**